lib/route.js

```diff
@@ -208,51 +228,57 @@ function buildRouting (options) {
       }
 
       if (prefixing === false) {
-      // run 'onRoute' hooks
+        // run 'onRoute' hooks
         for (const hook of this[kHooks].onRoute) {
-          try {
-            hook.call(this, opts)
-          } catch (error) {
-            done(error)
-            return
-          }
+          hook.call(this, opts)
         }
       }
 
+      const constraints = opts.constraints || {}
       const config = {
         ...opts.config,
         url,
         method: opts.method
       }
-      const constraints = opts.constraints || {}
-      if (opts.version) {
-        warning.emit('FSTDEP006')
-        constraints.version = opts.version
-      }
-
       const context = new Context(
         opts.schema,
         opts.handler.bind(this),
         this[kReply],
         this[kRequest],
         this[kContentTypeParser],
-        config,
-        opts.errorHandler || this[kErrorHandler],
-        opts.bodyLimit,
-        opts.logLevel,
-        opts.logSerializers,
-        opts.attachValidation,
-        this[kReplySerializerDefault],
-        opts.schemaErrorFormatter || this[kSchemaErrorFormatter]
+        config
       )
 
-      const headRouteExists = router.find('HEAD', path) != null
+      if (opts.version) {
+        warning.emit('FSTDEP006')
+        constraints.version = opts.version
+      }
+
+      const headRouteExists = opts.method === 'HEAD' && router.find('HEAD', url, constraints) != null
+
+      // TODO: improve validation for auto expose HEAD route before
+      // adding a new route
+
+      // There's a bug when ignoreTrailingSlash !== true
+      // See
+      /*
+      └── /prefix
+          └── / (GET)
+              / (HEAD)
+              └── / (GET)
+      **/
```

Comment:
This is a pretty interesting error popping up only when `ignoreTrailingSlash === true`, I didn't dig deep too much but was quite hard to detect. The third route comes out nowhere and the call stack is not really descriptive. I can take a look into it but not sure if it belongs directly to this PR as it happens before targeting to `next` branch.

Reply:
I don't understand the error you are describing. As it happens on `master` as well, open a fresh issue.

Reply:
Not sure if it happens on `main` but caused me problems when pointing to `next`.
What it does is create a third route out of nowhere, as described in the comment. A third route under `/prefix//` is made, only if `ignoreTrailingSlash = true`. I'll try to see if I can reproduce it on `main`

Reply:
It happens in `next` without this PR?

Reply:
Incredibly I was not able to reproduce it on `next` after a rebase. I'll try to figure out if stills reproducible or maybe was part of my changes before getting it stable